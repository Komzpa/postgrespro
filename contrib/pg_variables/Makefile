# contrib/pg_variables/Makefile

MODULE_big = pg_variables
OBJS = pg_variables.o pg_variables_record.o $(WIN32RES)

EXTENSION = pg_variables
DATA = pg_variables--1.0.sql
PGFILEDESC = "pg_variables - sessional variables"

REGRESS = pg_variables

ifdef USE_PGXS
PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
else
subdir = contrib/pg_variables
top_builddir = ../..
include $(top_builddir)/src/Makefile.global
include $(top_srcdir)/contrib/contrib-global.mk
endif
