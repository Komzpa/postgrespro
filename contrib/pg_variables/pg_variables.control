# pg_variables extension
comment = 'session variables with various types'
default_version = '1.0'
module_pathname = '$libdir/pg_variables'
relocatable = true
