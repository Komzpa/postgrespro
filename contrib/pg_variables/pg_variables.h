/*-------------------------------------------------------------------------
 *
 * pg_variables.c
 *	  exported definitions for pg_variables.c
 *
 * Copyright (c) 2015-2016, Postgres Professional
 *
 *-------------------------------------------------------------------------
 */
#ifndef __PG_VARIABLES_H__
#define __PG_VARIABLES_H__

#include "access/htup.h"
#include "access/tupdesc.h"
#include "datatype/timestamp.h"
#include "utils/date.h"
#include "utils/hsearch.h"
#include "utils/numeric.h"
#include "utils/jsonb.h"

/* initial number of packages hashes */
#define NUMPACKAGES 8
#define NUMVARIABLES 16

typedef struct HashPackageEntry
{
	char			name[NAMEDATALEN];
	HTAB		   *variablesHash;
	/* Memory context for package variables for easy memory release */
	MemoryContext	hctx;
} HashPackageEntry;

typedef struct RecordVar
{
	HTAB		   *rhash;
	TupleDesc		tupdesc;
	/* Memory context for records hash table for easy memory release */
	MemoryContext	hctx;
	/* Hash function info */
	FmgrInfo		hash_proc;
	/* Match function info */
	FmgrInfo		cmp_proc;
} RecordVar;

typedef struct ScalarVar
{
	Datum			value;
	bool			is_null;
	bool			typbyval;
	int16			typlen;
} ScalarVar;

typedef struct HashVariableEntry
{
	char			name[NAMEDATALEN];
	union
	{
		ScalarVar	scalar;
		RecordVar	record;
	}				value;

	Oid				typid;
} HashVariableEntry;

typedef struct HashRecordKey
{
	Datum		value;
	bool		is_null;
	/* Hash function info */
	FmgrInfo   *hash_proc;
	/* Match function info */
	FmgrInfo   *cmp_proc;
} HashRecordKey;

typedef struct HashRecordEntry
{
	HashRecordKey	key;
	HeapTuple		tuple;
} HashRecordEntry;

extern void init_attributes(HashVariableEntry* variable, TupleDesc tupdesc,
							MemoryContext topctx);
extern void check_attributes(HashVariableEntry *variable, TupleDesc tupdesc);
extern void check_record_key(HashVariableEntry *variable, Oid typid);

extern void insert_record(HashVariableEntry* variable,
						  HeapTupleHeader tupleHeader);
extern bool update_record(HashVariableEntry *variable,
						  HeapTupleHeader tupleHeader);
extern bool delete_record(HashVariableEntry* variable, Datum value,
						  bool is_null);
extern void clean_records(HashVariableEntry *variable);

#endif   /* __PG_VARIABLES_H__ */
