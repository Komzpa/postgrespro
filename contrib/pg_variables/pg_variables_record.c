/*-------------------------------------------------------------------------
 *
 * pg_variables_record.c
 *	  Functions to work with record types
 *
 * Copyright (c) 2015-2016, Postgres Professional
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/htup_details.h"
#include "catalog/pg_collation.h"
#include "catalog/pg_type.h"
#include "utils/builtins.h"
#include "utils/datum.h"
#include "utils/memutils.h"
#include "utils/typcache.h"

#include "pg_variables.h"

/*
 * Hash function for records.
 *
 * We use the element type's default hash opclass, and the default collation
 * if the type is collation-sensitive.
 */
static uint32
record_hash(const void *key, Size keysize)
{
	HashRecordKey		k = *((const HashRecordKey *) key);
	Datum				h;

	if (k.is_null)
		return 0;

	h = FunctionCall1Coll(k.hash_proc, DEFAULT_COLLATION_OID, k.value);
	return DatumGetUInt32(h);
}

/*
 * Matching function for records, to be used in hashtable lookups.
 */
static int
record_match(const void *key1, const void *key2, Size keysize)
{
	HashRecordKey		k1 = *((const HashRecordKey *) key1);
	HashRecordKey		k2 = *((const HashRecordKey *) key2);
	Datum				c;

	if (k1.is_null)
	{
		if (k2.is_null)
			return 0;			/* NULL "=" NULL */
		else
			return 1;			/* NULL ">" not-NULL */
	}
	else if (k2.is_null)
		return -1;				/* not-NULL "<" NULL */

	c = FunctionCall2Coll(k1.cmp_proc, DEFAULT_COLLATION_OID,
						  k1.value, k2.value);
	return DatumGetInt32(c);
}

void
init_attributes(HashVariableEntry *variable, TupleDesc tupdesc,
				MemoryContext topctx)
{
	HASHCTL		ctl;
	char		hash_name[BUFSIZ];
	MemoryContext		oldcxt;
	RecordVar		   *record;
	TypeCacheEntry	   *typentry;
	Oid					keyid;

	Assert(variable->typid == RECORDOID);

	sprintf(hash_name, "Records hash for variable \"%s\"", variable->name);

	record = &(variable->value.record);
	record->hctx = AllocSetContextCreate(topctx,
										 hash_name,
										 ALLOCSET_DEFAULT_MINSIZE,
										 ALLOCSET_DEFAULT_INITSIZE,
										 ALLOCSET_DEFAULT_MAXSIZE);
	oldcxt = MemoryContextSwitchTo(record->hctx);
	record->tupdesc = CreateTupleDescCopyConstr(tupdesc);

	/* Initialize hash table. */
	ctl.keysize = sizeof(HashRecordKey);
	ctl.entrysize = sizeof(HashRecordEntry);
	ctl.hcxt = record->hctx;
	ctl.hash = record_hash;
	ctl.match = record_match;

	record->rhash = hash_create(hash_name, NUMVARIABLES, &ctl,
										HASH_ELEM | HASH_CONTEXT |
										HASH_FUNCTION | HASH_COMPARE);

	/* Get hash and match functions for key type. */
	keyid = record->tupdesc->attrs[0]->atttypid;
	typentry = lookup_type_cache(keyid,
								 TYPECACHE_HASH_PROC_FINFO |
								 TYPECACHE_CMP_PROC_FINFO);

	if (!OidIsValid(typentry->hash_proc_finfo.fn_oid))
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_FUNCTION),
				 errmsg("could not identify a hash function for type %s",
						format_type_be(keyid))));

	if (!OidIsValid(typentry->cmp_proc_finfo.fn_oid))
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_FUNCTION),
				 errmsg("could not identify a matching function for type %s",
						format_type_be(keyid))));

	fmgr_info(typentry->hash_proc_finfo.fn_oid, &record->hash_proc);
	fmgr_info(typentry->cmp_proc_finfo.fn_oid, &record->cmp_proc);

	MemoryContextSwitchTo(oldcxt);
}

/*
 * New record structure should be the same as the first record.
 */
void
check_attributes(HashVariableEntry *variable, TupleDesc tupdesc)
{
	int			i;

	Assert(variable->typid == RECORDOID);

	/* First, check columns count. */
	if (variable->value.record.tupdesc->natts != tupdesc->natts)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("new record structure differs from variable \"%s\" "
						"structure", variable->name)));

	/* Second, check columns type. */
	for (i = 0; i < tupdesc->natts; i++)
	{
		Form_pg_attribute	attr1 = variable->value.record.tupdesc->attrs[i],
							attr2 = tupdesc->attrs[i];

		if ((attr1->atttypid != attr2->atttypid)
			|| (attr1->attndims != attr2->attndims)
			|| (attr1->atttypmod != attr2->atttypmod))
			ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("new record structure differs from variable \"%s\" "
						"structure", variable->name)));
	}
}

/*
 * Check record key type. If not same then throw a error.
 */
void
check_record_key(HashVariableEntry *variable, Oid typid)
{
	Assert(variable->typid == RECORDOID);

	if (variable->value.record.tupdesc->attrs[0]->atttypid != typid)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("requested value type differs from variable \"%s\" "
						"key type", variable->name)));
}

/*
 * Insert a new record. New record key should be unique in the variable.
 */
void
insert_record(HashVariableEntry *variable, HeapTupleHeader tupleHeader)
{
	TupleDesc			tupdesc;
	HeapTuple			tuple;
	int					tuple_len;
	Datum				value;
	bool				isnull;
	RecordVar		   *record;
	HashRecordKey		k;
	HashRecordEntry	   *item;
	bool				found;
	MemoryContext		oldcxt;

	Assert(variable->typid == RECORDOID);

	record = &(variable->value.record);

	oldcxt = MemoryContextSwitchTo(record->hctx);

	tupdesc = record->tupdesc;

	/* Build a HeapTuple control structure */
	tuple_len = HeapTupleHeaderGetDatumLength(tupleHeader);

	tuple = (HeapTuple) palloc(HEAPTUPLESIZE + tuple_len);
	tuple->t_len = tuple_len;
	ItemPointerSetInvalid(&(tuple->t_self));
	tuple->t_tableOid = InvalidOid;
	tuple->t_data = (HeapTupleHeader) ((char *) tuple + HEAPTUPLESIZE);
	memcpy((char *) tuple->t_data, (char *) tupleHeader, tuple_len);

	/* Inserting a new record */
	value = fastgetattr(tuple, 1, tupdesc, &isnull);
	/* First, check if there is a record with same key */
	k.value = value;
	k.is_null = isnull;
	k.hash_proc = &record->hash_proc;
	k.cmp_proc = &record->cmp_proc;

	item = (HashRecordEntry *) hash_search(record->rhash, &k,
										   HASH_ENTER, &found);
	if (found)
	{
		heap_freetuple(tuple);
		MemoryContextSwitchTo(oldcxt);
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("there is a record in the variable \"%s\" with same "
						"key", variable->name)));
	}
	/* Second, insert a new record */
	item->tuple = tuple;

	MemoryContextSwitchTo(oldcxt);
}

/*
 * Insert a record. New record key should be unique in the variable.
 */
bool
update_record(HashVariableEntry* variable, HeapTupleHeader tupleHeader)
{
	TupleDesc			tupdesc;
	HeapTuple			tuple;
	int					tuple_len;
	Datum				value;
	bool				isnull;
	RecordVar		   *record;
	HashRecordKey		k;
	HashRecordEntry	   *item;
	bool				found;
	MemoryContext		oldcxt;

	Assert(variable->typid == RECORDOID);

	record = &(variable->value.record);

	oldcxt = MemoryContextSwitchTo(record->hctx);

	tupdesc = record->tupdesc;

	/* Build a HeapTuple control structure */
	tuple_len = HeapTupleHeaderGetDatumLength(tupleHeader);

	tuple = (HeapTuple) palloc(HEAPTUPLESIZE + tuple_len);
	tuple->t_len = tuple_len;
	ItemPointerSetInvalid(&(tuple->t_self));
	tuple->t_tableOid = InvalidOid;
	tuple->t_data = (HeapTupleHeader) ((char *) tuple + HEAPTUPLESIZE);
	memcpy((char *) tuple->t_data, (char *) tupleHeader, tuple_len);

	/* Update a record */
	value = fastgetattr(tuple, 1, tupdesc, &isnull);
	k.value = value;
	k.is_null = isnull;
	k.hash_proc = &record->hash_proc;
	k.cmp_proc = &record->cmp_proc;

	item = (HashRecordEntry *) hash_search(record->rhash, &k,
										   HASH_FIND, &found);
	if (!found)
	{
		heap_freetuple(tuple);
		MemoryContextSwitchTo(oldcxt);
		return false;
	}

	item->tuple = tuple;

	MemoryContextSwitchTo(oldcxt);
	return true;
}

bool
delete_record(HashVariableEntry *variable, Datum value, bool is_null)
{
	HashRecordKey		k;
	HashRecordEntry	   *item;
	bool				found;
	RecordVar		   *record;

	Assert(variable->typid == RECORDOID);

	record = &(variable->value.record);

	/* Delete a record */
	k.value = value;
	k.is_null = is_null;
	k.hash_proc = &record->hash_proc;
	k.cmp_proc = &record->cmp_proc;

	item = (HashRecordEntry *) hash_search(record->rhash, &k,
										   HASH_REMOVE, &found);
	if (found)
		heap_freetuple(item->tuple);

	return found;
}

/*
 * Free allocated space for records hash table and datums array.
 */
void
clean_records(HashVariableEntry *variable)
{
	Assert(variable->typid == RECORDOID);

	hash_destroy(variable->value.record.rhash);
	FreeTupleDesc(variable->value.record.tupdesc);

	/* All records will be freed */
	MemoryContextDelete(variable->value.record.hctx);
}
