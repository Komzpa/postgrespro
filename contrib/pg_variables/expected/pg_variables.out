CREATE EXTENSION pg_variables;
-- Integer variables
SELECT pgv_get_int('vars', 'int1');
ERROR:  unrecognized package "vars"
SELECT pgv_get_int('vars', 'int1', false);
 pgv_get_int 
-------------
            
(1 row)

SELECT pgv_set_int('vars', 'int1', 101);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_set_int('vars', 'int2', 102);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
 pgv_get_int 
-------------
         101
(1 row)

SELECT pgv_get_int('vars', 'int2');
 pgv_get_int 
-------------
         102
(1 row)

SELECT pgv_set_int('vars', 'int1', 103);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
 pgv_get_int 
-------------
         103
(1 row)

SELECT pgv_get_int('vars', 'int3');
ERROR:  unrecognized variable "int3"
SELECT pgv_get_int('vars', 'int3', false);
 pgv_get_int 
-------------
            
(1 row)

SELECT pgv_exists('vars', 'int3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'int1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_set_int('vars', 'intNULL', NULL);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'intNULL');
 pgv_get_int 
-------------
            
(1 row)

-- Text variables
SELECT pgv_set_text('vars', 'str1', 's101');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_set_text('vars', 'int1', 's101');
ERROR:  variable "int1" requires "integer" value
SELECT pgv_set_int('vars', 'str1', 101);
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_text('vars', 'str2', 's102');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'str1');
 pgv_get_text 
--------------
 s101
(1 row)

SELECT pgv_get_text('vars', 'str2');
 pgv_get_text 
--------------
 s102
(1 row)

SELECT pgv_set_text('vars', 'str1', 's103');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'str1');
 pgv_get_text 
--------------
 s103
(1 row)

SELECT pgv_get_text('vars', 'str3');
ERROR:  unrecognized variable "str3"
SELECT pgv_get_text('vars', 'str3', false);
 pgv_get_text 
--------------
 
(1 row)

SELECT pgv_exists('vars', 'str3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'str1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_text('vars', 'int1');
ERROR:  variable "int1" requires "integer" value
SELECT pgv_get_int('vars', 'str1');
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_text('vars', 'strNULL', NULL);
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'strNULL');
 pgv_get_text 
--------------
 
(1 row)

-- Numeric variables
SELECT pgv_set_numeric('vars', 'num1', 1.01);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_set_numeric('vars', 'num2', 1.02);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_set_numeric('vars', 'str1', 1.01);
ERROR:  variable "str1" requires "text" value
SELECT pgv_get_numeric('vars', 'num1');
 pgv_get_numeric 
-----------------
            1.01
(1 row)

SELECT pgv_get_numeric('vars', 'num2');
 pgv_get_numeric 
-----------------
            1.02
(1 row)

SELECT pgv_set_numeric('vars', 'num1', 1.03);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_get_numeric('vars', 'num1');
 pgv_get_numeric 
-----------------
            1.03
(1 row)

SELECT pgv_get_numeric('vars', 'num3');
ERROR:  unrecognized variable "num3"
SELECT pgv_get_numeric('vars', 'num3', false);
 pgv_get_numeric 
-----------------
                
(1 row)

SELECT pgv_exists('vars', 'num3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'num1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_numeric('vars', 'str1');
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_numeric('vars', 'numNULL', NULL);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_get_numeric('vars', 'numNULL');
 pgv_get_numeric 
-----------------
                
(1 row)

SET timezone = 'Europe/Moscow';
-- Timestamp variables
SELECT pgv_set_timestamp('vars', 'ts1', '2016-03-30 10:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_set_timestamp('vars', 'ts2', '2016-03-30 11:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_set_timestamp('vars', 'num1', '2016-03-30 12:00:00');
ERROR:  variable "num1" requires "numeric" value
SELECT pgv_get_timestamp('vars', 'ts1');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 10:00:00 2016
(1 row)

SELECT pgv_get_timestamp('vars', 'ts2');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 11:00:00 2016
(1 row)

SELECT pgv_set_timestamp('vars', 'ts1', '2016-03-30 12:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_get_timestamp('vars', 'ts1');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 12:00:00 2016
(1 row)

SELECT pgv_get_timestamp('vars', 'ts3');
ERROR:  unrecognized variable "ts3"
SELECT pgv_get_timestamp('vars', 'ts3', false);
 pgv_get_timestamp 
-------------------
 
(1 row)

SELECT pgv_exists('vars', 'ts3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'ts1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_timestamp('vars', 'num1');
ERROR:  variable "num1" requires "numeric" value
SELECT pgv_set_timestamp('vars', 'tsNULL', NULL);
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_get_timestamp('vars', 'tsNULL');
 pgv_get_timestamp 
-------------------
 
(1 row)

-- TimestampTZ variables
SELECT pgv_set_timestamptz('vars', 'tstz1', '2016-03-30 10:00:00 GMT+01');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_set_timestamptz('vars', 'tstz2', '2016-03-30 11:00:00 GMT+02');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_set_timestamptz('vars', 'ts1', '2016-03-30 12:00:00 GMT+03');
ERROR:  variable "ts1" requires "timestamp without time zone" value
SELECT pgv_get_timestamptz('vars', 'tstz1');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 14:00:00 2016 MSK
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz2');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 16:00:00 2016 MSK
(1 row)

SELECT pgv_set_timestamptz('vars', 'tstz1', '2016-03-30 12:00:00 GMT+01');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz1');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 16:00:00 2016 MSK
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz3');
ERROR:  unrecognized variable "tstz3"
SELECT pgv_get_timestamptz('vars', 'tstz3', false);
 pgv_get_timestamptz 
---------------------
 
(1 row)

SELECT pgv_exists('vars', 'tstz3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'tstz1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_timestamptz('vars', 'ts1');
ERROR:  variable "ts1" requires "timestamp without time zone" value
SELECT pgv_set_timestamptz('vars', 'tstzNULL', NULL);
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstzNULL');
 pgv_get_timestamptz 
---------------------
 
(1 row)

-- Date variables
SELECT pgv_set_date('vars', 'd1', '2016-03-29');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_set_date('vars', 'd2', '2016-03-30');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_set_date('vars', 'tstz1', '2016-04-01');
ERROR:  variable "tstz1" requires "timestamp with time zone" value
SELECT pgv_get_date('vars', 'd1');
 pgv_get_date 
--------------
 03-29-2016
(1 row)

SELECT pgv_get_date('vars', 'd2');
 pgv_get_date 
--------------
 03-30-2016
(1 row)

SELECT pgv_set_date('vars', 'd1', '2016-04-02');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_get_date('vars', 'd1');
 pgv_get_date 
--------------
 04-02-2016
(1 row)

SELECT pgv_get_date('vars', 'd3');
ERROR:  unrecognized variable "d3"
SELECT pgv_get_date('vars', 'd3', false);
 pgv_get_date 
--------------
 
(1 row)

SELECT pgv_exists('vars', 'd3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'd1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_date('vars', 'tstz1');
ERROR:  variable "tstz1" requires "timestamp with time zone" value
SELECT pgv_set_date('vars', 'dNULL', NULL);
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_get_date('vars', 'dNULL');
 pgv_get_date 
--------------
 
(1 row)

-- Jsonb variables
SELECT pgv_set_jsonb('vars2', 'j1', '[1, 2, "foo", null]');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_set_jsonb('vars2', 'j2', '{"bar": "baz", "balance": 7.77, "active": false}');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_set_jsonb('vars', 'd1', '[1, 2, "foo", null]');
ERROR:  variable "d1" requires "date" value
SELECT pgv_get_jsonb('vars2', 'j1');
    pgv_get_jsonb    
---------------------
 [1, 2, "foo", null]
(1 row)

SELECT pgv_get_jsonb('vars2', 'j2');
                  pgv_get_jsonb                   
--------------------------------------------------
 {"bar": "baz", "active": false, "balance": 7.77}
(1 row)

SELECT pgv_set_jsonb('vars2', 'j1', '{"foo": [true, "bar"], "tags": {"a": 1, "b": null}}');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_get_jsonb('vars2', 'j1');
                    pgv_get_jsonb                    
-----------------------------------------------------
 {"foo": [true, "bar"], "tags": {"a": 1, "b": null}}
(1 row)

SELECT pgv_get_jsonb('vars2', 'j3');
ERROR:  unrecognized variable "j3"
SELECT pgv_get_jsonb('vars2', 'j3', false);
 pgv_get_jsonb 
---------------
 
(1 row)

SELECT pgv_exists('vars2', 'j3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars2', 'j1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_jsonb('vars', 'd1');
ERROR:  variable "d1" requires "date" value
SELECT pgv_set_jsonb('vars', 'jNULL', NULL);
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_get_jsonb('vars', 'jNULL');
 pgv_get_jsonb 
---------------
 
(1 row)

-- Record variables
CREATE TABLE tab (id int, t varchar);
INSERT INTO tab VALUES (0, 'str00'), (1, 'str11'), (2, NULL), (NULL, 'strNULL');
SELECT pgv_insert('vars3', 'r1', tab) FROM tab;
 pgv_insert 
------------
 
 
 
 
(4 rows)

SELECT pgv_insert('vars2', 'j1', tab) FROM tab;
ERROR:  variable "j1" requires "jsonb" value
SELECT pgv_insert('vars3', 'r1', tab) FROM tab;
ERROR:  there is a record in the variable "r1" with same key
SELECT pgv_insert('vars3', 'r1', row(1, 'str1', 'str2'));
ERROR:  new record structure differs from variable "r1" structure
SELECT pgv_insert('vars3', 'r1', row(1, 1));
ERROR:  new record structure differs from variable "r1" structure
SELECT pgv_insert('vars3', 'r1', row('str1', 'str1'));
ERROR:  new record structure differs from variable "r1" structure
SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (1,str11)
 (2,)
 (0,str00)
(4 rows)

SELECT pgv_select('vars3', 'r1', 1);
 pgv_select 
------------
 (1,str11)
(1 row)

SELECT pgv_select('vars3', 'r1', 0);
 pgv_select 
------------
 (0,str00)
(1 row)

SELECT pgv_select('vars3', 'r1', NULL::int);
 pgv_select 
------------
 (,strNULL)
(1 row)

SELECT pgv_select('vars3', 'r1', ARRAY[1, 0, NULL]);
 pgv_select 
------------
 (1,str11)
 (0,str00)
 (,strNULL)
(3 rows)

UPDATE tab SET t = 'str33' WHERE id = 1;
SELECT pgv_update('vars3', 'r1', tab) FROM tab;
 pgv_update 
------------
 t
 t
 t
 t
(4 rows)

SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (1,str33)
 (2,)
 (0,str00)
(4 rows)

SELECT pgv_delete('vars3', 'r1', 1);
 pgv_delete 
------------
 t
(1 row)

SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (2,)
 (0,str00)
(3 rows)

SELECT pgv_delete('vars3', 'r1', 100);
 pgv_delete 
------------
 f
(1 row)

SELECT pgv_select('vars3', 'r3');
ERROR:  unrecognized variable "r3"
SELECT pgv_exists('vars3', 'r3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars3', 'r1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_select('vars2', 'j1');
ERROR:  variable "j1" requires "jsonb" value
-- Manipulate variables
SELECT * FROM pgv_list() order by package, name;
 package |   name   
---------+----------
 vars    | d1
 vars    | d2
 vars    | dNULL
 vars    | int1
 vars    | int2
 vars    | intNULL
 vars    | jNULL
 vars    | num1
 vars    | num2
 vars    | numNULL
 vars    | str1
 vars    | str2
 vars    | strNULL
 vars    | ts1
 vars    | ts2
 vars    | tsNULL
 vars    | tstz1
 vars    | tstz2
 vars    | tstzNULL
 vars2   | j1
 vars2   | j2
 vars3   | r1
(22 rows)

SELECT pgv_remove('vars', 'int3');
ERROR:  unrecognized variable "int3"
SELECT pgv_remove('vars', 'int1');
 pgv_remove 
------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
ERROR:  unrecognized variable "int1"
SELECT pgv_remove('vars2');
 pgv_remove 
------------
 
(1 row)

SELECT pgv_get_jsonb('vars2', 'j1');
ERROR:  unrecognized package "vars2"
SELECT * FROM pgv_list() order by package, name;
 package |   name   
---------+----------
 vars    | d1
 vars    | d2
 vars    | dNULL
 vars    | int2
 vars    | intNULL
 vars    | jNULL
 vars    | num1
 vars    | num2
 vars    | numNULL
 vars    | str1
 vars    | str2
 vars    | strNULL
 vars    | ts1
 vars    | ts2
 vars    | tsNULL
 vars    | tstz1
 vars    | tstz2
 vars    | tstzNULL
 vars3   | r1
(19 rows)

SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

SELECT * FROM pgv_list() order by package, name;
 package | name 
---------+------
(0 rows)

